drivers/flash/flash_gd32f4xx.c

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (c) 2022 BrainCo Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <gd32_fmc.h>
+
+#include "flash_gd32.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(flash_gd32);
+
+#define FLASH_GD32F4XX_WRITE_ERR (FMC_STAT_PGMERR | FMC_STAT_PGSERR | FMC_STAT_WPERR)
+#define FLASH_GD32F4XX_ERASE_ERR FMC_STAT_OPERR
+
+/**
+ * @brief GD32F4xx FMC use sector as the minimal operate unit.
+ * For GD32F425x, GD32F427x and GD32F470x, page are also supported, page size
+ * is 4KB.
+ */
+#ifdef CONFIG_FLASH_PAGE_LAYOUT
+#if (PRE_KB(512) == FLASH_GD32_SIZE)
+static const struct flash_pages_layout flash_gd32f4xx_layout[] = {
+	/* GD32F4XXxE */
+	{.pages_count = 4, .pages_size = KB(16)},
+	{.pages_count = 1, .pages_size = KB(64)},
+	{.pages_count = 3, .pages_size = KB(128)},
+};
+#elif (PRE_KB(1024) == FLASH_GD32_SIZE)
+static const struct flash_pages_layout flash_gd32f4xx_layout[] = {
+	/*GD32F4XXxG */
+	{.pages_count = 4, .pages_size = KB(16)},
+	{.pages_count = 1, .pages_size = KB(64)},
+	{.pages_count = 7, .pages_size = KB(128)},
+};
+#elif (PRE_KB(2048) == FLASH_GD32_SIZE)
+static const struct flash_pages_layout flash_gd32f4xx_layout[] = {
+	/* GD32F4XXxI */
+	{.pages_count = 4, .pages_size = KB(16)},
+	{.pages_count = 1, .pages_size = KB(64)},
+	{.pages_count = 7, .pages_size = KB(128)},
+	{.pages_count = 4, .pages_size = KB(16)},
+	{.pages_count = 1, .pages_size = KB(64)},
+	{.pages_count = 7, .pages_size = KB(128)},
+};
+#elif (PRE_KB(3072) == FLASH_GD32_SIZE)
+static const struct flash_pages_layout flash_gd32f4xx_layout[] = {
+	/* GD32F4XXxK */
+	{.pages_count = 4, .pages_size = KB(16)},
+	{.pages_count = 1, .pages_size = KB(64)},
+	{.pages_count = 7, .pages_size = KB(128)},
+	{.pages_count = 4, .pages_size = KB(16)},
+	{.pages_count = 1, .pages_size = KB(64)},
+	{.pages_count = 7, .pages_size = KB(128)},
+	{.pages_count = 4, .pages_size = KB(256)},
+};
+#else
+#error "Unknown FMC layout for GD32F4xx series."
+#endif
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+
+/* SN bits in FMC_CTL are not continues value, use below table to mapping it. */
```

Comment:
`/* SN bits in FMC_CTL are not continue values, use table below to map them. */`

Reply:
Thanks for your advice! done.